File: 백준/Silver/14281. 볼록 수열/볼록 수열.py
import sys
input = sys.stdin.readline

n = int(input())
arr = list(map(int, input().split()))
cnt = 0

def back(n, arr):
    global cnt
    is_convex = True
    for i in range(1, n-1):
        if arr[i-1]+arr[i+1] < 2*arr[i]:
            tmp = arr[i]- (arr[i-1] + arr[i+1])//2
            arr[i] -= tmp
            cnt += tmp
            is_convex = False
    return cnt, is_convex

while True:
    cnt, is_convex = back(n, arr)
    if is_convex: 
        break
  
print(cnt)

File: 백준/Silver/14281. 볼록 수열/볼록 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        BufferedReader br = new BufferedReader(new FileReader("./input.txt"));
        int n = Integer.parseInt(br.readLine());
        int[] numbers = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        ConvexNumbers convexNumbers = new ConvexNumbers(n, numbers);
        convexNumbers.activate();

    }

    static class ConvexNumbers {
        private final int n;
        private int[] numbers;
        private boolean isEnd;
        private long cnt;

        public ConvexNumbers(int n, int[] numbers) {
            this.n = n;
            this.numbers = numbers;
            this.isEnd = false;
            this.cnt = 0;
        }

        public void activate() {
            while (!isEnd) {
                isEnd = true;
                for (int i = 1; i < n - 1; i++) {
                    if (!isConvex(i)) {
                        int dif = numbers[i] - (numbers[i - 1] + numbers[i + 1]) / 2;
                        cnt += dif;
                        numbers[i] -= dif;
                        isEnd = false;
                    }
                }
            }
            System.out.println(cnt);
        }

        private boolean isConvex(int idx) {
            return numbers[idx - 1] + numbers[idx + 1] >= 2 * numbers[idx];
        }
    }
}